src/back/space/src/api/invitation.cpp

```diff
@@ -0,0 +1,202 @@
+			return Delete(req, res);
+		default:
+			throw std::runtime_error("Unsupported");
+			break;
+		}
+	} catch(const errors::Unauthorized& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kUnauthorized);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const errors::BadRequest& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kBadRequest);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const errors::NotFound& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kNotFound);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const errors::NotModified& e) {
+		req.SetResponseStatus(server::http::HttpStatus::kNotModified);
+		res["err"] = e.what();
+		return res.ExtractValue();
+	} catch(const std::exception& e) {
+		LOG_WARNING() << "Fail to process invitation handle with method: "
+			<< req.GetMethodStr() << " err: " << e.what();
+		res["err"] = "Fail to process invitation";
+		req.SetResponseStatus(server::http::HttpStatus::kInternalServerError);
+	}
+	return res.ExtractValue();
+}
+
+formats::json::Value Invitation::GetList(
+	const server::http::HttpRequest& req,
+	formats::json::ValueBuilder& res) const
+{
+	auto paging = parsePaging(req);
+	const auto list = _s.GetInvitationList(paging.start, paging.limit);
+	res["list"] = list.items;
+	res["total"] = list.total;
+
+	return res.ExtractValue();
+}
+
+formats::json::Value Invitation::Post(
+	const server::http::HttpRequest& req,
+	const formats::json::Value& body,
+	formats::json::ValueBuilder& res) const
+{
+	const auto& creatorId = req.GetHeader(headers::kUserId);
+	if (creatorId.empty())
+		throw errors::Unauthorized{};
+
+	if (req.HasArg("link")) {
+		const auto link = req.GetArg("link");
+		if (link.empty())
+			throw errors::BadRequest{"Empty linkId"};
+
+		if (_s.IsLinkExpired(link))
+			throw errors::BadRequest{"Link expired"};
+
+		_s.InviteByLink(creatorId, link);
+
+		req.SetResponseStatus(server::http::HttpStatus::kCreated);
+		return res.ExtractValue();
+	}
+
+	auto invitation = body.As<model::SpaceInvitation>();
+
+	if (invitation.spaceId.is_nil() || invitation.userId.empty())
+		throw errors::BadRequest{"Params must be set"};
+
+	if (!_s.ValidateRole(invitation.role))
+		throw errors::BadRequest{"Wrong role"};
+
+	_s.Invite(creatorId, invitation.spaceId, invitation.userId, invitation.role);
+
+	req.SetResponseStatus(server::http::HttpStatus::kCreated);
+
+	return res.ExtractValue();
+}
+
+formats::json::Value Invitation::ChangeRole(
+	const server::http::HttpRequest& req,
+	const formats::json::Value& body,
+	formats::json::ValueBuilder& res) const
+{
+	const auto& id = req.GetArg("id");
+	if (id.empty())
+		throw errors::BadRequest{"Param id must be set"};
+
+	int iId;
+
+	try {
+		iId = std::atoi(id.c_str());
+	} catch(const std::exception& e) {
+		throw errors::BadRequest("Id param must be valid");
+	}
+
+	if (iId < 0)
+		throw errors::BadRequest{"Id param must be valid"};
```

Comment:
- Извини, с именем функции я ошибся, не atoi, а std::stoi
- Вынеси пожалуйста весь этот кусок куда то в shared

```suggestion
	const auto id = parsePositiveInt(req, "id");
```

Где-то в shared в .cpp
```c++
int parsePositiveInt(const server::http::HttpRequest& req, const std::string& key)
{
	try {
		const int value = std::stoi(req.GetArg(key));
		if (value < 0)
			throw std::runtime_error{"must be greater then -1"};
		return value;
	} catch(const std::exception& e) {
		const auto msg = fmt::format("Param {} parse err: {}", key, e.what());
		throw errors::BadRequest(msg);
	}
	return 0;
}
```

Reply:
Готово (везде).